### ISSUE TITLE
Install Plugin using script fails due to 404

### ISSUE THREAD

**Issue by rahulsom** [jenkinsci/docker#470]
### Docker commands that you execute
I had this in my `Dockerfile`.
```
FROM jenkinsci/jenkins:lts
RUN /usr/local/bin/install-plugins.sh build-timeout
```

I ran 
```
docker build .
```

### Actual result
The logs say
```
Failed to download plugin: build-timeout or build-timeout-plugin
```
Opening the url on my browser causes a redirect to 
http://mirrors.jenkins-ci.org/plugins/build-timeout-plugin/latest/build-timeout-plugin.hpi

That results in 404

### Expected outcome
The plugin would be installed in my docker image

### Have you tried a non-dockerized Jenkins and get the expected outcome?
No

### Output of `docker version`
```
Client:
 Version:      17.03.1-ce
 API version:  1.27
 Go version:   go1.7.5
 Git commit:   c6d412e
 Built:        Tue Mar 28 00:40:02 2017
 OS/Arch:      darwin/amd64

Server:
 Version:      17.03.1-ce
 API version:  1.27 (minimum version 1.12)
 Go version:   go1.7.5
 Git commit:   c6d412e
 Built:        Fri Mar 24 00:00:50 2017
 OS/Arch:      linux/amd64
 Experimental: true
```

### Other relevant information
Once the UI is launched, installing plugins works. I believe it hits a different url than the `install-plugins.sh`.

**Comment by carlossg** [comment]
this is probably a temporary issue

**Comment by rtyler** [comment]
No, that file really doesn't exist. http://mirrors.jenkins-ci.org/plugins/build-timeout/latest/build-timeout.hpi exists however

**Comment by carlossg** [comment]
it should also try http://mirrors.jenkins-ci.org/plugins/build-timeout/latest/build-timeout.hpi which does exist

**Comment by tanandy** [comment]
i had the same problem and solved it with my own retry mechanism:

i think the curl retry mechanism doesnt handle properly all "transient errors" (maybe because of using -o or -O)

so implementing your own retry mechanism fix the problem:

it fails because if the first plugin download url fails (because of network :) , it will try the second url (with "-plugin" this time that returns 404 since the first one was the good url).

so if your first download succeed it will not try a bad url

example:

install-plugins.sh:

add this function
``` bash

retry() {
    local max_attempts="$1"; shift
    local attempt=1
    local cmd="$@"

    until $cmd
    do
        if (( attempt == max_attempts ))
        then
            echo "Attempt $attempt failed and there are no more attempts left!"
            return 1
        else
            echo "Attempt $attempt failed! Trying again in $attempt seconds..."
            sleep $(( attempt++ ))
        fi
    done
}
```
then replace with this line
```bash
retry ${CURL_RETRY:-5} curl --connect-timeout "${CURL_CONNECTION_TIMEOUT:-20}" --retry "${CURL_RETRY:-5}" --retry-delay "${CURL_RETRY_DELAY:-0}" --retry-max-time "${CURL_RETRY_MAX_TIME:-60}" -s -f -L "$url" -o "$jpi"
```

**Comment by carlossg** [comment]
A 404 is not a transient error. If the build fails now then you'll need to re-run the build later on, but not in the install-plugins script

**Comment by tanandy** [comment]
we dont want to rerun build a lot of times until my network works well (we have bad network so sometimes we need 4 5 times before build succeed)
so retry solution fixed my problem

**Comment by carlossg** [comment]
See #636

**Comment by sanga** [comment]
Sweet. Thank you good sir
